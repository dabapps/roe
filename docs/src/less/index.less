@import '../../../src/less/index.less';

@body-background: @grey-lighter;

h1,
h2 {
  color: @secondary;
}

.container {
  box-shadow: @shadow-hard;
}

.display-columns {
  .column {
    background-color: @grey-lightest;
    border: @border-base;
  }
}
