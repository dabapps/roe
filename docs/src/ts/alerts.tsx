import * as React from 'react';

import {
  Alert,
  CodeBlock,
  Column,
  DabIpsum,
  Row,
  Section
} from '../../../src/ts';

const Alerts = () => (
  <Section>
    <Row>
      <Column>
        <h2>
          Alerts
        </h2>

        <Alert>
          <DabIpsum count={2} />
        </Alert>
        <Alert className="info">
          <DabIpsum count={2} />
        </Alert>
        <Alert className="success">
          <DabIpsum count={2} />
        </Alert>
        <Alert className="warning">
          <DabIpsum count={2} />
        </Alert>
        <Alert className="error">
          <DabIpsum count={2} />
        </Alert>
      </Column>
    </Row>
    <Row>
      <Column>
        <h3>
          Code
        </h3>
        <CodeBlock language="javascript" name="JSX">
          {`
            <Alert>
              <DabIpsum count={2} />
            </Alert>
            <Alert className="info">
              <DabIpsum count={2} />
            </Alert>
            <Alert className="success">
              <DabIpsum count={2} />
            </Alert>
            <Alert className="warning">
              <DabIpsum count={2} />
            </Alert>
            <Alert className="error">
              <DabIpsum count={2} />
            </Alert>
          `}
        </CodeBlock>
      </Column>
    </Row>
  </Section>
);

export default Alerts;
