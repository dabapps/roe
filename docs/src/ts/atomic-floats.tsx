import * as React from 'react';

import {
  CodeBlock,
  Column,
  Row,
  Section
} from '../../../src/ts';

const AtomicFloats = () => (
  <Section>
    <Row>
      <Column>
        <h2>
          Atomic Float Classes
        </h2>
      </Column>
    </Row>
    <Row>
      <Column className="display-columns">
        <h3>
          Demo
        </h3>
        <Row>
          <Column xs={6} className="xs-float-right sm-float-left md-float-right">
            float-right sm-float-left md-float-right
          </Column>
          <Column xs={6} className="xs-float-left sm-float-right md-float-left">
            float-left sm-float-right md-float-left
          </Column>
        </Row>
        <p>
          Note: float-left is an alias for xs-float-left, and float-right an alias for xs-float-right
        </p>
      </Column>
    </Row>
    <Row>
      <Column>
        <h3>
          Code
        </h3>
        <CodeBlock language="javascript">
          {`
            <Row>
              <Column xs={6} className="xs-float-right sm-float-left md-float-right">
                float-right sm-float-left md-float-right
              </Column>
              <Column xs={6} className="xs-float-left sm-float-right md-float-left">
                float-left sm-float-right md-float-left
              </Column>
            </Row>
          `}
        </CodeBlock>
      </Column>
    </Row>
  </Section>
);

export default AtomicFloats;
