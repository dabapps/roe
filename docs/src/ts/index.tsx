import * as React from 'react';
import * as ReactDOM from 'react-dom';

import './livereload';

import {
  Column,
  Container,
  Row
} from '../../../src/ts';

import { Buttons } from './buttons';
import { CodeBlocks } from './code-blocks';
import { Grid } from './grid';
import { Inputs } from './inputs';
import { Text } from './text';

const packageJson = require( '../../../package.json'); // tslint:disable-line:no-var-requires

class App extends React.Component<void, void> {
  public render () {
    return (
      <Container>
        <Row>
          <Column>
            <h1>
              Roe - DabApps' Project Development Kit (v{packageJson.version})
            </h1>
            <p>
              <strong>
                A Collection of React Components for Project Development
              </strong>
            </p>
          </Column>
        </Row>
        <Buttons />
        <CodeBlocks />
        <Grid />
        <Inputs />
        <Text />
      </Container>
    );
  }
}

ReactDOM.render(<App />, document.getElementById('app'));
