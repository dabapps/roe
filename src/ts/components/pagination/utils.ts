import { MAX_BUTTONS, LEFT_BUTTONS, RIGHT_BUTTONS } from './constants';

export const getIsDots = (
  totalPages: number,
  index: number,
  page: number
): boolean => {
  const hasMorePagesThanButtons = totalPages > MAX_BUTTONS;

  return (
    hasMorePagesThanButtons &&
    ((index === 1 && page > 2) ||
      (index === MAX_BUTTONS - 2 && page < totalPages - 1))
  );
};

export const getButtonType = (
  totalPages: number,
  page: number,
  index: number,
  currentPage: number
): string | undefined => {
  if (currentPage === page) {
    return 'selected';
  }
  if (getIsDots(totalPages, index, page)) {
    return 'dots';
  }

  return undefined;
};

export const getPageToGoTo = (
  totalPages: number,
  page: number,
  index: number
): number => {
  if (totalPages > MAX_BUTTONS && index === 0 && page > 1) {
    return 1;
  } else if (
    totalPages > MAX_BUTTONS &&
    index === MAX_BUTTONS - 1 &&
    page < totalPages
  ) {
    return totalPages;
  }

  return page;
};

const getStart = (totalPages: number, currentPage: number) => {
  return Math.max(
    Math.min(
      currentPage - LEFT_BUTTONS,
      totalPages - LEFT_BUTTONS - RIGHT_BUTTONS
    ),
    1
  );
};

const getEnd = (totalPages: number, start: number) =>
  Math.min(start + MAX_BUTTONS, totalPages + 1);

const getRange = (itemCount: number, pageSize: number, pageCount: number) => {
  const remainder = itemCount % pageSize;

  if (remainder === 0 && pageCount < MAX_BUTTONS) {
    return Math.floor(pageCount);
  }
  if (remainder !== 0 && pageCount < MAX_BUTTONS) {
    return Math.floor(pageCount) + 1;
  }

  return MAX_BUTTONS;
};

export const paginationSeries = (
  totalPages: number,
  pageCount: number,
  itemCount: number,
  pageSize: number,
  currentPage: number
): readonly number[] => {
  const start = getStart(totalPages, currentPage);
  const end = getEnd(totalPages, start);
  const range = getRange(itemCount, pageSize, pageCount);

  return [...Array(range)].map((_item, index) =>
    Math.floor(start + index * ((end - start) / range))
  );
};
