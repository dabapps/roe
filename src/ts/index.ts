export { default as Alert } from './components/misc/alert';
export { default as AppRoot } from './components/app/root';
export { default as Anchor } from './components/misc/anchor';
export { default as Badge } from './components/misc/badge';
export { default as Banner } from './components/banners/banner';
export { default as Button } from './components/forms/button';
export { default as CodeBlock } from './components/misc/code-block';
export { default as Collapse } from './components/misc/collapse';
export { default as Column } from './components/grid/column';
export { default as Container } from './components/grid/container';
export { default as ContentBox } from './components/content/content-box';
export { default as ContentBoxHeader } from './components/content/content-box-header';
export { default as ContentBoxFooter } from './components/content/content-box-footer';
export { default as CookieBanner } from './components/banners/cookie-banner';
export { CookieBannerRenderProps } from './components/banners/cookie-banner';
export { default as DabIpsum } from './components/prototyping/dab-ipsum';
export { default as Footer } from './components/navigation/footer';
export { default as FormGroup } from './components/forms/form-group';
export { default as Highlight } from './components/misc/highlight';
export { default as InputGroup } from './components/forms/input-group';
export { default as InputGroupAddon } from './components/forms/input-group-addon';
export { default as InputWithPrefixSuffix } from './components/precomposed/input-with-prefix-suffix';
export { default as Modal } from './components/modals/modal';
export { default as ModalBody } from './components/modals/modal-body';
export { default as ModalCloseIcon } from './components/modals/modal-close-icon';
export { default as ModalFooter } from './components/modals/modal-footer';
export { default as ModalHeader } from './components/modals/modal-header';
export { default as ModalRenderer } from './components/modals/modal-renderer';
export { default as Nav } from './components/navigation/nav';
export { default as NavBar } from './components/navigation/nav-bar';
export { default as NavItem } from './components/navigation/nav-item';
export { default as Row } from './components/grid/row';
export { default as Section } from './components/content/section';
export { default as SideBar } from './components/navigation/side-bar';
export { default as SpacedGroup } from './components/misc/spaced-group';
export { default as SpeechBubble } from './components/misc/speech-bubble';
export { default as Tabs } from './components/tabs/tabs';
export { default as Tab } from './components/tabs/tab';
export { default as Table } from './components/tables/table';
export { default as TableBody } from './components/tables/table-body';
export { default as TableCell } from './components/tables/table-cell';
export { default as TableHead } from './components/tables/table-head';
export { default as TableHeader } from './components/tables/table-header';
export { default as TableRow } from './components/tables/table-row';
export { default as Well } from './components/misc/well';
