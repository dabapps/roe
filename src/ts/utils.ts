import * as randomSeed from 'random-seed';

const MATCHES_INITIAL_INDENTATION = /^([^\S\n]*)\S/;
const MATCHES_BLANK_FIRST_LINE = /^\s*\n/;
const MATCHES_BLANK_LAST_LINE = /\n\s*$/;

export function formatCode (code: string) {
  const codeWithoutLeadingOrTrailingEmptyLines = code
    .replace(MATCHES_BLANK_FIRST_LINE, '')
    .replace(MATCHES_BLANK_LAST_LINE, '');

  const initialIndentation: RegExpExecArray | null =
    MATCHES_INITIAL_INDENTATION.exec(codeWithoutLeadingOrTrailingEmptyLines);

  return initialIndentation ?
    codeWithoutLeadingOrTrailingEmptyLines.replace(new RegExp(`^${initialIndentation[1]}`, 'gm'), '') :
    codeWithoutLeadingOrTrailingEmptyLines;
}

const MATCHES_AMPERSAND = /&/gi;
const MATCHES_NON_WORD_CHARACTERS = /[\W_]+/gi;
const MATCHES_LEADING_AND_TRAILING_HYPHENS = /(^-+|-+$)/gi;

export const getHref = (children?: React.ReactNode, href?: string): string | undefined => {
  if (href) {
    return href;
  }

  if (typeof children !== 'string') {
    return undefined;
  }

  return children
    .replace(MATCHES_AMPERSAND, '-and-')
    .replace(MATCHES_NON_WORD_CHARACTERS, '-')
    .replace(MATCHES_LEADING_AND_TRAILING_HYPHENS, '')
    .toLowerCase();
}

let rand = randomSeed.create('dabapps');

export const resetRandomSeed = () => {
  rand = randomSeed.create('dabapps');
}

export const generateIpsum = (words: string[]) => {
  const ipsum = Array.apply(null, new Array(15)).map(() => (
    words[Math.floor(rand.range(words.length))]
  )).join(' ');

  return ipsum.charAt(0).toUpperCase() + ipsum.substring(1) + '.';
}
