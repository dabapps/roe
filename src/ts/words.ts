export const WORDS = [
  'academia',
  'agile',
  'angular',
  'apprentice',
  'apps',
  'brighton',
  'business',
  'careers',
  'cats',
  'client',
  'coffee',
  'community',
  'css',
  'dabapps',
  'design',
  'designers',
  'development',
  'developers',
  'digital',
  'django',
  'documentation',
  'education',
  'equality',
  'events',
  'framework',
  'green',
  'innovation',
  'ionic',
  'ios',
  'java',
  'javascript',
  'kittens',
  'llamas',
  'mobile',
  'native',
  'networking',
  'news',
  'objective-c',
  'open-source',
  'partnerships',
  'product',
  'projects',
  'prototyping',
  'python',
  'react',
  'react-native',
  'recognition',
  'rest',
  'service',
  'smashing',
  'tea',
  'teaching',
  'technical',
  'testimonial',
  'testing',
  'toolkit',
  'training',
  'typescript',
  'web',
];
