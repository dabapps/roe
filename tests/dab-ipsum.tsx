import * as React from 'react';
import * as renderer from 'react-test-renderer';
import * as enzyme from 'enzyme';

import { DabIpsum } from '../src/ts';
import { generateIpsum, resetRandomSeed } from '../src/ts/utils';
import { WORDS } from '../src/ts/words';

describe('DabIpsum', () => {
  beforeEach(() => {
    resetRandomSeed();
  });

  it('should render some paragraphs by default', () => {
    const tree = renderer.create(<DabIpsum />);

    expect(tree).toMatchSnapshot();
  });

  it('should allow setting how many items to render', () => {
    const tree = renderer.create(<DabIpsum count={3} />);

    expect(tree).toMatchSnapshot();
  });

  it('should allow rendering an unordered list', () => {
    const tree = renderer.create(<DabIpsum component="ul" />);

    expect(tree).toMatchSnapshot();
  });

  it('should allow rendering an ordered list', () => {
    const tree = renderer.create(<DabIpsum component="ol" />);

    expect(tree).toMatchSnapshot();
  });

  it('should allow rendering plain text (in a span)', () => {
    const tree = renderer.create(<DabIpsum component="text" />);

    expect(tree).toMatchSnapshot();
  });

  it('should generate some ipsum', () => {
    const ipsum = generateIpsum(WORDS);

    expect(ipsum).toMatchSnapshot();
  });

  it('should only update if type or count changed', () => {
    const instance = enzyme.mount(<DabIpsum component="text" count={1} />);

    expect(instance).toMatchSnapshot();

    instance.setProps({ component: 'text', count: 1 });

    expect(instance).toMatchSnapshot();

    instance.setProps({ component: 'p', count: 1 });

    expect(instance).toMatchSnapshot();

    instance.setProps({ component: 'p', count: 1 });

    expect(instance).toMatchSnapshot();

    instance.setProps({ component: 'p', count: 2 });

    expect(instance).toMatchSnapshot();
  });
});
