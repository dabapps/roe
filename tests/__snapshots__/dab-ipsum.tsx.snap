// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DabIpsum should allow rendering an ordered list 1`] = `
<ol>
  <li>
    Teaching kittens coffee agile developers apps training product projects toolkit events tea careers cats javascript.
  </li>
  <li>
    Objective-c testimonial academia projects designers coffee toolkit apprentice business product technical documentation react-native tea designers.
  </li>
  <li>
    Python tea angular client development service news testimonial academia prototyping development technical apps events web.
  </li>
  <li>
    Development innovation technical ios service networking testimonial technical web technical cats agile django service documentation.
  </li>
  <li>
    Design toolkit networking mobile service recognition documentation ionic product llamas testing react-native product networking python.
  </li>
</ol>
`;

exports[`DabIpsum should allow rendering an unordered list 1`] = `
<ul>
  <li>
    Teaching kittens coffee agile developers apps training product projects toolkit events tea careers cats javascript.
  </li>
  <li>
    Objective-c testimonial academia projects designers coffee toolkit apprentice business product technical documentation react-native tea designers.
  </li>
  <li>
    Python tea angular client development service news testimonial academia prototyping development technical apps events web.
  </li>
  <li>
    Development innovation technical ios service networking testimonial technical web technical cats agile django service documentation.
  </li>
  <li>
    Design toolkit networking mobile service recognition documentation ionic product llamas testing react-native product networking python.
  </li>
</ul>
`;

exports[`DabIpsum should allow rendering plain text (in a span) 1`] = `
<span>
  Teaching kittens coffee agile developers apps training product projects toolkit events tea careers cats javascript.
</span>
`;

exports[`DabIpsum should allow setting how many items to render 1`] = `
<div>
  <p>
    Teaching kittens coffee agile developers apps training product projects toolkit events tea careers cats javascript.
  </p>
  <p>
    Objective-c testimonial academia projects designers coffee toolkit apprentice business product technical documentation react-native tea designers.
  </p>
  <p>
    Python tea angular client development service news testimonial academia prototyping development technical apps events web.
  </p>
</div>
`;

exports[`DabIpsum should generate some ipsum 1`] = `"Teaching kittens coffee agile developers apps training product projects toolkit events tea careers cats javascript."`;

exports[`DabIpsum should only update if type or count changed 1`] = `
<Memo(DabIpsum)
  component="text"
  count={1}
>
  <span
    key="0"
  >
    Teaching kittens coffee agile developers apps training product projects toolkit events tea careers cats javascript.
  </span>
</Memo(DabIpsum)>
`;

exports[`DabIpsum should only update if type or count changed 2`] = `
<Memo(DabIpsum)
  component="text"
  count={1}
>
  <span
    key="0"
  >
    Teaching kittens coffee agile developers apps training product projects toolkit events tea careers cats javascript.
  </span>
</Memo(DabIpsum)>
`;

exports[`DabIpsum should only update if type or count changed 3`] = `
<Memo(DabIpsum)
  component="p"
  count={1}
>
  <div>
    <p
      key="0"
    >
      Objective-c testimonial academia projects designers coffee toolkit apprentice business product technical documentation react-native tea designers.
    </p>
  </div>
</Memo(DabIpsum)>
`;

exports[`DabIpsum should only update if type or count changed 4`] = `
<Memo(DabIpsum)
  component="p"
  count={1}
>
  <div>
    <p
      key="0"
    >
      Objective-c testimonial academia projects designers coffee toolkit apprentice business product technical documentation react-native tea designers.
    </p>
  </div>
</Memo(DabIpsum)>
`;

exports[`DabIpsum should only update if type or count changed 5`] = `
<Memo(DabIpsum)
  component="p"
  count={2}
>
  <div>
    <p
      key="0"
    >
      Python tea angular client development service news testimonial academia prototyping development technical apps events web.
    </p>
    <p
      key="1"
    >
      Development innovation technical ios service networking testimonial technical web technical cats agile django service documentation.
    </p>
  </div>
</Memo(DabIpsum)>
`;

exports[`DabIpsum should render some paragraphs by default 1`] = `
<div>
  <p>
    Teaching kittens coffee agile developers apps training product projects toolkit events tea careers cats javascript.
  </p>
  <p>
    Objective-c testimonial academia projects designers coffee toolkit apprentice business product technical documentation react-native tea designers.
  </p>
  <p>
    Python tea angular client development service news testimonial academia prototyping development technical apps events web.
  </p>
  <p>
    Development innovation technical ios service networking testimonial technical web technical cats agile django service documentation.
  </p>
  <p>
    Design toolkit networking mobile service recognition documentation ionic product llamas testing react-native product networking python.
  </p>
</div>
`;
